hibernate-core/src/main/java/org/hibernate/event/internal/DefaultDeleteEventListener.java

```diff
@@ -131,6 +131,7 @@ public void onDelete(DeleteEvent event, Set transientEntities) throws HibernateE
 					persister,
 					false
 			);
+			persister.afterReassociate( entity, source );
```

Comment:
After adding a detached entity to the `PersistenceContext` (just above the lines shown above), `EntityPersister#afterReassociate` needs to be called.  If the entity is enhanced for lazy loading, then `EntityPersister#afterReassociate` sets the `Session` in an enhanced entity's `LazyAttributeLoadingInterceptor`.

If that's not done, cascade-delete won't work properly.